dd-sdk-android/src/test/kotlin/com/datadog/android/DatadogEventListenerFactoryTest.kt

```diff
@@ -0,0 +1,131 @@
+/*
+ * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
+ * This product includes software developed at Datadog (https://www.datadoghq.com/).
+ * Copyright 2016-Present Datadog, Inc.
+ */
+
+package com.datadog.android
+
+import com.datadog.android.utils.forge.Configurator
+import com.nhaarman.mockitokotlin2.verify
+import fr.xgouchet.elmyr.annotation.LongForgery
+import fr.xgouchet.elmyr.annotation.RegexForgery
+import fr.xgouchet.elmyr.junit5.ForgeConfiguration
+import fr.xgouchet.elmyr.junit5.ForgeExtension
+import java.io.IOException
+import java.net.InetSocketAddress
+import java.net.Proxy
+import okhttp3.Call
+import okhttp3.Connection
+import okhttp3.EventListener
+import okhttp3.Protocol
+import okhttp3.Request
+import okhttp3.Response
+import org.junit.jupiter.api.BeforeEach
+import org.junit.jupiter.api.Test
+import org.junit.jupiter.api.extension.ExtendWith
+import org.junit.jupiter.api.extension.Extensions
+import org.mockito.Answers
+import org.mockito.Mock
+import org.mockito.junit.jupiter.MockitoExtension
+import org.mockito.junit.jupiter.MockitoSettings
+import org.mockito.quality.Strictness
+
+@Extensions(
+    ExtendWith(MockitoExtension::class),
+    ExtendWith(ForgeExtension::class)
+)
+@MockitoSettings(strictness = Strictness.LENIENT)
+@ForgeConfiguration(Configurator::class)
+class DatadogEventListenerFactoryTest {
+    @Mock(answer = Answers.RETURNS_MOCKS)
+    lateinit var mockCall: Call
+
+    @RegexForgery("[a-z]+\\.[a-z]{3}")
+    lateinit var fakeDomain: String
+
+    lateinit var fakeRequest: Request
+    lateinit var fakeResponse: Response
+
+    val fakeIOException: IOException = IOException("Fake error")
+
+    @Mock
+    lateinit var mockConnecion: Connection
+
+    @Mock
+    lateinit var mockDelegatedListener: EventListener
+
+    lateinit var testedFactory: DatadogEventListener.Factory
+
+    @LongForgery(min = 1L)
+    var fakeByteCount: Long = 1L
+
+    @BeforeEach
+    fun `set up`() {
+        fakeRequest = Request.Builder().get().url("https://$fakeDomain/").build()
+        fakeResponse = Response.Builder()
+            .request(fakeRequest)
+            .protocol(Protocol.HTTP_2)
+            .code(200)
+            .message("lorem ipsum dolor sit amet…")
+            .build()
+
+        val delegatedFactory = EventListener.Factory { mockDelegatedListener }
+        testedFactory = DatadogEventListener.Factory(delegatedFactory)
+    }
+
+    @Test
+    fun `𝕄 delegate are called 𝕎 all methods`() {
+        // Given
+        val testedListener = DatadogEventListener.Factory(testedFactory).create(mockCall)
+        val address = InetSocketAddress(0)
+
+        // When
+        testedListener.callStart(mockCall)
```

Comment:
those assertions/test methods should be moved in the `DatadogEventListenerTest` class as follows:

```
fun `M delegate W responseHeadersStart()`{
     // GIVEN
     testedListener = DatadogEventListener(fakeKey, mockDelegate)
    
    // WHEN
    testedListener.responseHeadersStart(mockCall)
   
   // THEN
   verify(mockDelegate).responseHeaderStart(mockCall)
   verifyNoMoreInteractions(mockDelegate)
}
```
There should be a similar test method for each of the `EventListener` implemented method.
The main reason behind this approach is to have each test method testing one single key feature in the SDK, this way it is easier for us to see/identify the issue whenever something goes wrong in our CI server.

Reply:
My first implementation was in `DatadogEventListenerTest`, but then I changed in order to make sure delegates are working via the factory. 

> The main reason behind this approach is to have each test method testing one single key feature in the SDK, this way it is easier for us to see/identify the issue whenever something goes wrong in our CI server.

In this case, would be more practical to write a dynamic test that collects all methods and make sure no one is missed.

Reply:
I am not sure how this dynamic test would look like but if it is going to have a dedicated test method for delegated method inside the EventListener that's the way. I know it is a lot of boiler code but it is better if you want to have relevant status messages from your CI server.

Reply:
Regarding the factory, its only responsibility is to build that instance (DatadogEventListener in this case). You don't have its business logic, the most you have to do is to test that it really contains that delegate. The business logic is going to be tested by the `DatadogEventListenerTest` class.